### ISSUE TITLE
guard-for-in won't show error if logic contains unrelated IF statement

### ISSUE THREAD

**Issue by TheOptimisticFactory** [eslint/eslint#14129]
### Environment ####

* **ESLint Version:**: `v7.19.0`
* **Node Version:**: `v14.15.5`
* **npm Version:**: `v6.14.9`
* **Operating System:**: `Ubuntu budgie v20.10`
* **Parser**: `@babel/eslint-parser`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  env: {
    es2020: true,
    node: true,
    mocha: true
  },
  extends: [
    "eslint:recommended",
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:node/recommended",
    "plugin:security/recommended",
  ],
  plugins: [
    "chai-expect",
    "filenames",
    "mocha",
    "promise",
    "sort-destructure-keys",
    "sort-keys-fix",
  ],
  parser: '@babel/eslint-parser',
  parserOptions: {
    sourceType: 'module',
    ecmaVersion: 2020,
    allowImportExportEverywhere: false,
    codeFrame: false,
  },
  rules: {
    /* ======================= */
    /* ||  Possible Errors  || */
    /* ======================= */
    "no-buffer-constructor": "error",
    "no-cond-assign": "error",
    "no-console": "error",
    "no-constant-condition": "error",
    "no-control-regex": "error",
    "no-debugger": "error",
    "no-dupe-args": "error",
    "no-dupe-keys": "error",
    "no-duplicate-case": "error",
    "no-empty-character-class": "error",
    "no-empty": "error",
    "no-ex-assign": "error",
    "no-extra-boolean-cast": "error",
    //"no-extra-parens": "error",
    "no-extra-semi": "error",
    "no-func-assign": "error",
    "no-inner-declarations": "error",
    "no-invalid-regexp": "error",
    "no-irregular-whitespace": "error",
    "no-obj-calls": "error",
    "no-prototype-builtins": "error",
    "no-regex-spaces": "error",
    "no-sparse-arrays": "error",
    "no-template-curly-in-string": "error",
    "no-unexpected-multiline": "error",
    "no-unreachable": "error",
    "no-unsafe-finally": "error",
    "no-unsafe-negation": "error",
    "use-isnan": "error",
    /*
    "valid-jsdoc": ["error", {
      "matchDescription": ".+",
      "requireReturn": false,
      "prefer": {
        "arg": "param",
        "argument": "param",
        "class": "constructor",
        "return": "returns",
        "virtual": "abstract"
      },
      "preferType": {
        "boolean": "Boolean",
        "number": "Number",
        "object": "Object",
        "string": "String"
      }
    }],
    */
    "valid-typeof": "error",

    /* ===================== */
    /* ||  Best Practices || */
    /* ===================== */
    "accessor-pairs": "error",
    "array-callback-return": "error",
    "block-scoped-var": "error",
    "class-methods-use-this": "off",
    "complexity": ["error", 20],
    "consistent-return": "error",
    "curly": ["error", "all"],
    "default-case": "error",
    "dot-location": ["error", "property"],
    "dot-notation": "error",
    "eqeqeq": "error",
    "guard-for-in": "error",
    "no-alert": "error",
    "no-caller": "error",
    // no-case-declarations
    "no-div-regex": "error",
    "no-else-return": "error",
    // no-empty-function
    "no-empty-pattern": "error",
    "no-eq-null": "error",
    "no-eval": "error",
    "no-extend-native": "error",
    "no-extra-bind": "error",
    // no-extra-label
    "no-fallthrough": "error",
    "no-floating-decimal": "error",
    "no-global-assign": "error",
    "no-implicit-coercion": "error",
    "no-implicit-globals": "error",
    "no-implied-eval": "error",
    "no-invalid-this": "error",
    "no-iterator": "error",
    "no-labels": "error",
    "no-lone-blocks": "error",
    "no-loop-func": "error",
    "no-magic-numbers": ["error", {
      "ignoreArrayIndexes": true,
      "ignore": [0, 1],
      "enforceConst": true
    }],
    "no-multi-spaces": "error",
    "no-multi-str": "error",
    "no-new-func": "error",
    "no-new-wrappers": "error",
    "no-new": "error",
    "no-octal-escape": "error",
    "no-octal": "error",
    // no-param-reassign
    "no-proto": "error",
    "no-redeclare": "error",
    // COULD BE USEFUL :
    // no-restricted-properties
    "no-return-assign": "error",
    "no-return-await": "error",
    "no-script-url": "error",
    "no-self-assign": "error",
    "no-self-compare": "error",
    "no-sequences": "error",
    "no-throw-literal": "error",
    "no-unmodified-loop-condition": "error",
    // no-unused-labels
    "no-useless-call": "error",
    "no-useless-concat": "error",
    "no-useless-escape": "error",
    "no-useless-return": "error",
    "no-void": "error",
    // no-warning-comments
    "no-with": "error",
    "prefer-promise-reject-errors": "error",
    // radix
    "require-await": "off",
    "vars-on-top": "error",
    // wrap-iife
    "yoda": "error",

    /* ===================== */
    /* ||    Variables    || */
    /* ===================== */
    // init-declarations
    "no-catch-shadow": "error",
    "no-delete-var": "error",
    // no-label-var
    // COULD BE USEFUL :
    // no-restricted-globals
    "no-shadow-restricted-names": "error",
    // Could be useful :
    "no-shadow": "error",
    "no-undef-init": "error",
    "no-undef": "error",
    // no-undefined
    "no-unused-vars": ["error", {
      args: "after-used",
      argsIgnorePattern: "^(queryInterface|Sequelize|DataTypes|toolkit)$",
      ignoreRestSiblings: true,
    }],
    "no-use-before-define": ["error", {
      "functions": false
    }],

    /* ============================ */
    /* ||  Node.js and CommonJS  || */
    /* ============================ */
    "callback-return": ["error", ["callback", "cb", "next"]],
    "global-require": "error",
    "handle-callback-err": "error",
    // no-mixed-requires
    "no-new-require": "error",
    "no-path-concat": "error",
    "no-process-env": "error",
    "no-process-exit": "error",
    // no-restricted-modules
    "no-sync": "error",

    /* ======================== */
    /* ||  Stylistic Issues  || */
    /* ======================== */
    "array-bracket-spacing": ["error", "always", {
      "objectsInArrays": false,
      "arraysInArrays": false
    }],
    "block-spacing": "error",
    "brace-style": ["error", "1tbs", {
      "allowSingleLine": true
    }],
    // capitalized-comments
    "comma-dangle": ["error", "always-multiline"],
    "comma-spacing": "error",
    "comma-style": "error",
    "computed-property-spacing": "error",
    "consistent-this": ["error", "self"],
    "eol-last": "error",
    "func-call-spacing": "error",
    "func-name-matching": "error",
    "func-names": "error",
    "func-style": ["error", "declaration", {
      "allowArrowFunctions": true
    }],
    // COULD BE USEFUL:
    // id-blacklist
    "id-length": ["error", {
      "exceptions": ["x", "y", "i", "j"]
    }],
    // id-match
    "indent": ["error", 2],
    "jsx-quotes": "error",
    "key-spacing": "error",
    "keyword-spacing": "error",
    // line-comment-position
    // Too many error currently :
    //"linebreak-style": "error",
    "lines-around-comment": ["error", {
      "beforeLineComment": true,
      "allowBlockStart": true,
      "allowObjectStart": true,
      "allowArrayStart": true
    }],
    "lines-around-directive": "error",
    "max-depth": ["error", 4],
    "max-len": ["error", 150, {
      "ignoreComments": true
    }],
    "max-lines": ["error", {
      "skipBlankLines": true,
      "skipComments": true
    }],
    "max-nested-callbacks": ["error", {
      "max": 4
    }],
    "max-params": ["warn", 3],
    "max-statements-per-line": ["error", {
      "max": 1
    }],
    // max-statements
    // multiline-ternary
    "new-parens": "error",
    "padding-line-between-statements": [
      "error",
      { blankLine: "always", prev: ["const", "let", "var"], next: "*" },
      { blankLine: "any",    prev: ["const", "let", "var"], next: ["const", "let", "var"] },
      { blankLine: "always", prev: ["block-like"], next: "*" },
    ],
    "newline-before-return": "error",
    "newline-per-chained-call": "error",
    "no-array-constructor": "error",
    "no-bitwise": "error",
    "no-continue": "error",
    "no-inline-comments": "error",
    "no-lonely-if": "error",
    "no-mixed-operators": "error",
    "no-mixed-spaces-and-tabs": "error",
    "no-multi-assign": "error",
    "no-multiple-empty-lines": ["error", {
      "max": 1,
      "maxEOF": 1,
      "maxBOF": 1
    }],
    // no-negated-condition
    "no-nested-ternary": "error",
    "no-new-object": "error",
    // no-plusplus
    // no-restricted-syntax
    "no-tabs": "error",
    // no-ternary
    "no-trailing-spaces": ["error", {
      "skipBlankLines": true
    }],
    // no-underscore-dangle
    "no-unneeded-ternary": "error",
    "no-whitespace-before-property": "error",
    "object-curly-newline": ["error", {
      "ObjectExpression": {
        "minProperties": 1
      },
      "ObjectPattern": {
        "multiline": true
      }
    }],
    "object-curly-spacing": ["error", "always", {
      "arraysInObjects": false,
      "objectsInObjects": false
    }],
    "object-property-newline": "error",
    "one-var-declaration-per-line": ["error", "always"],
    "one-var": ["error", "never"],
    // operator-assignment
    "operator-linebreak": ["error", "after", {
      "overrides": {
        "?": "before",
        ":": "before"
      }
    }],
    // padded-blocks
    "quote-props": ["error", "as-needed"],
    "quotes": ["error", "single"],
    /*
    "require-jsdoc": ["warn", {
      "require": {
        "FunctionDeclaration": true,
        "MethodDefinition": true,
        "ClassDeclaration": true
      }
    }],
    */
    "semi-spacing": "error",
    "semi-style": ["error", "last"],
    "semi": "error",
    "sort-keys-fix/sort-keys-fix": ['error', 'asc', {
      caseSensitive: false,
    }],
    "sort-vars": "error",
    "space-before-blocks": "error",
    "space-before-function-paren": ["error", {
      "anonymous": "never",
      "named": "never",
      "asyncArrow": "always"
    }],
    "space-in-parens": "error",
    "space-infix-ops": "error",
    "space-unary-ops": ["error", {
      "words": true,
      "nonwords": false
    }],
    "spaced-comment": ["error", "always"],
    "template-tag-spacing": "error",
    "unicode-bom": "error",
    // wrap-regex

    /* ===================== */
    /* ||   ECMAScript 6  || */
    /* ===================== */
    "arrow-body-style": ["error", "as-needed"],
    "arrow-parens": ["error", "as-needed"],
    "arrow-spacing": "error",
    "constructor-super": "error",
    "generator-star-spacing": "error",
    "no-class-assign": "error",
    "no-confusing-arrow": ["error", {
      "allowParens": true
    }],
    "no-const-assign": "error",
    "no-dupe-class-members": "error",
    "no-duplicate-imports": "error",
    "no-new-symbol": "error",
    // no-restricted-imports
    "no-this-before-super": "error",
    "no-useless-computed-key": "error",
    "no-useless-constructor": "error",
    "no-useless-rename": "error",
    "no-var": "error",
    "object-shorthand": "error",
    "prefer-arrow-callback": "error",
    "prefer-const": "error",
    "prefer-numeric-literals": "error",
    "prefer-rest-params": "error",
    "prefer-spread": "error",
    "prefer-template": "error",
    "require-yield": "error",
    "rest-spread-spacing": "error",
    "sort-imports": [ "error",  {
      ignoreCase: true,
    }],
    "symbol-description": "error",
    "template-curly-spacing": "error",
    "yield-star-spacing": "error",
    "require-atomic-updates": "off",

    /* ===================== */
    /* ||      MOCHA      || */
    /* ===================== */
    "mocha/handle-done-callback": "error",
    "mocha/no-exclusive-tests": "error",
    "mocha/no-global-tests": "error",
    "mocha/no-hooks-for-single-case": "error",
    "mocha/no-identical-title": "error",
    "mocha/no-nested-tests": "error",
    "mocha/no-pending-tests": "error",
    "mocha/no-sibling-hooks": "error",
    "mocha/no-skipped-tests": "error",
    "mocha/no-top-level-hooks": "error",

    /* ==================== */
    /* ||      CHAI      || */
    /* ==================== */
    "chai-expect/no-inner-compare": "error",
    "chai-expect/missing-assertion": "error",
    "chai-expect/terminating-properties": "error",

    /* ========================= */
    /* ||      FILENAMES      || */
    /* ========================= */
    "filenames/match-exported": "off",

    /* ======================== */
    /* ||      PROMISES      || */
    /* ======================== */
    "promise/always-return": "error",
    "promise/no-return-wrap": "error",
    "promise/param-names": "error",
    "promise/catch-or-return": "error",
    "promise/no-nesting": "warn",
    "promise/no-promise-in-callback": "warn",
    "promise/no-callback-in-promise": "warn",
    "promise/avoid-new": "warn",
    "promise/prefer-await-to-then": "warn",

    /* ======================== */
    /* ||        NODE        || */
    /* ======================== */
    "node/no-extraneous-require": "off",
    "node/no-deprecated-api": "warn",
    "node/no-unpublished-require": "off",

    /* ============================ */
    /* ||        SECURITY        || */
    /* ============================ */
    "security/detect-object-injection": "off",
    "security/detect-non-literal-fs-filename": "off",

    /* ========================================= */
    /* ||        SORT-DESTRUCTURE-KEYS        || */
    /* ========================================= */
    "sort-destructure-keys/sort-destructure-keys": [ "error", {
      caseSensitive: false,
    }]
  },
  globals: {
    Child: true
  },
  overrides: [

    /* ======================================== */
    /* ||      OVERRIDES - CONFIG FILES      || */
    /* ======================================== */
    {
      files: "config/**/*.js",
      rules: {
        "no-process-env": "off",
      },
    },

    /* ====================================== */
    /* ||      OVERRIDES - TEST FILES      || */
    /* ====================================== */
    {
      files: "**/*.test.js",
      rules: {
        "camelcase": "off",
        "class-methods-use-this": "off",
        "max-len": "off",
        "no-magic-numbers": "off",
        "no-unused-expressions": "off",
        "require-jsdoc": "off",
      },
    },

    /* ======================================== */
    /* ||      OVERRIDES - MODELS FILES      || */
    /* ======================================== */
    {
      files: "services/database/models/*.js",
      rules: {
        "sort-keys-fix/sort-keys-fix": "off"
      }
    },

    /* =========================================== */
    /* ||      OVERRIDES - MIGRATION FILES      || */
    /* =========================================== */
    {
      files: "services/database/migrations/scripts/*.js",
      rules: {
        "sort-keys-fix/sort-keys-fix": "off"
      }
    },

    /* ====================================== */
    /* ||      OVERRIDES - TEST FILES      || */
    /* ====================================== */
    {
      files: "**/*.test.js",
      rules: {
        "camelcase": "off",
        "class-methods-use-this": "off",
        "max-len": "off",
        "max-lines": "off",
        "no-magic-numbers": "off",
        "no-unused-expressions": "off",
        "node/no-unpublished-require": "off",
        "require-jsdoc": "off",
        "sort-keys-fix/sort-keys-fix": "off"
      }
    },

    /* ====================================== */
    /* ||      OVERRIDES - TEST FILES      || */
    /* ====================================== */
    {
      files: "test/fixtures/*.js",
      rules: {
        "max-len": "off",
        "max-lines": "off",
        "sort-keys-fix/sort-keys-fix": "off"
      }
    },

    /* ======================================================= */
    /* ||      OVERRIDES - OPENAPI SPECIFICATION FILES      || */
    /* ======================================================= */
    {
      files: "routes/*.specifications.js",
      rules: {
        "max-len": "off",
        "max-lines": "off",
        "no-magic-numbers": "off",
      }
    }, {
      files: "routes/specifications/{parameters,responses,schemas,tags}.js",
      rules: {
        "max-len": "off",
        "max-lines": "off",
        "no-magic-numbers": "off",
      }
    },
  ],
}

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
'use strict';

/*eslint guard-for-in: "error"*/

function demo(element = {}) {
  const unrelatedNumber = 42;
  const doSomething = () => {};
  
  // Does not trigger any `guard-for-in` error
  for (const property in element) {
    if (Math.random() === unrelatedNumber) {
      doSomething(property);
    }
  }

  // Does not trigger any `guard-for-in` error
  for (const property in element) {
    if (element[property] === null || element[property] === undefined) {
      doSomething[property];
    }
  }

  // Does not trigger any `guard-for-in` error
  for (const property in element) {
    if (JSON.stringify(true) === 'true') {
      doSomething(property);
    }
  }

  // The last one does trigger a `guard-for-in` error
  for (const property in element) {
    doSomething(property);
  }

  return element;
}

module.exports = demo;
```

```bash
cross-env FORCE_COLOR=1 eslint .
```

**What did you expect to happen?**

I would expect all 4 of these for-in loops to throw a **guard-for-in** error as none of these truly filter unwanted properties from the prototype.

**What actually happened? Please copy-paste the actual, raw output from ESLint.**

![image](https://user-images.githubusercontent.com/2607260/108838031-6223e180-75d3-11eb-9a08-d5d49be3192f.png)

Only the 4th loop triggers an error despite all 4 possibly accessing properties that are inherited through the prototype chain.
```js
24:3  error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
```

**Steps to reproduce this issue:**

1. Create a for-in loop if any IF condition that isn't any listed in the `Examples of current code` in the rule documentation: https://eslint.org/docs/rules/guard-for-in
2. Run eslint

**Comment by mdjermanovic** [comment]
Hi @TheOptimisticFactory, thanks for the issue!

The rule works as intended:

> This rule is aimed at preventing unexpected behavior that could arise from using a for in loop without filtering the results in the loop. As such, it will warn when for in loops do not filter their results with an if statement.

If you have an idea on how could we improve the rule to ensure that the filtering is indeed based on a condition that checks whether or not the property is an own property of the object, please open a Rule change request.

I'm closing this issue since it isn't a bug.

**Comment by TheOptimisticFactory** [comment]
Shouldn't the rule check for an exact match between `Object.prototype.hasOwnProperty.call` or `{}.hasOwnProperty.call` or any whitelisted function passed as an option of said rule?

None of the above IF statements of my example for-in loops have anything to due to filtering prototype properties.

**Comment by mdjermanovic** [comment]
This was discussed in the original issue about this rule, and that's issue #1 :)

If you think that the rule can be improved now, please feel free to open a [rule change request](https://github.com/eslint/eslint/blob/master/.github/ISSUE_TEMPLATE/RULE_CHANGE.md) issue with the proposal.
